### PR TITLE
Github Liquid tag: Fix error when repository has no README + refactor

### LINKED ISSUE #6382: GitHub tag fails if no readme
**Describe the bug**
When I try to create a GitHub repo embed of a repo without a readme, the tag errors with a 404 (see image below).

**To Reproduce**

1. Find a repo with no readme (I used b3u/chordy, but subsequently added a readme)
2. Create a new post or comment containing `{% github :user/:repo %}`
3. Preview, and see the error

**Expected behavior**

Possible solutions include displaying "readme not found" in place of the readme itself, and simply defaulting to the `no-readme` option.

**Screenshots**
image

**Desktop (please complete the following information):**
- OS: Ubuntu 18.04 (Elementary OS v5.1)
- Browser: chrome
- Version: 80

**Additional context**

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Bug Fix

## Description

Our Github liquid tag raised an error when we tried to display a repository with no README unless the user specified the "no-readme" option. This PR fixes that.

Note: I found the code relatively hard to follow, so I refactored it quite a bit in the process. I'll add some more comments inline regarding the what/why.

## Related Tickets & Documents

Closes #6382 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

Before:

Screen Shot 2020-03-11 at 11 43 58

After:

Screen Shot 2020-03-11 at 11 44 21

Repositories with READMEs work like before:

Screen Shot 2020-03-11 at 12 19 42

## Added tests?

- [X] yes

## Added to documentation?

- [X] no documentation needed

### CODE REVIEW

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 1]
```diff
@@ -4,38 +4,35 @@ class GithubTag
   class GithubReadmeTag
     PARTIAL = "liquids/github_readme".freeze
 
+    attr_reader :client, :content, :options, :readme_html
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: Just added some readers to use throughout the file.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 2]
```diff
@@ -4,38 +4,35 @@ class GithubTag
...
+
     def initialize(link)
-      @link = parse_link(link)
+      parsed_link = parse_link(link)
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: IMHO a better name. Also, this doesn't need to be an instance variable since it gets passed into all methods relying on it.

I also moved link validation + raising of an error into `parse_link`. Before we parsed the link, then validated it in `get_content`. Now we exit as early as possible, avoiding all unncessary work.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 3]
```diff
@@ -4,38 +4,35 @@ class GithubTag
...
       @options = parse_options(link)
-      @content = get_content(@link)
+      @client = Octokit::Client.new(access_token: token)
+      @content = client.repository(parsed_link)
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: After splitting responsibilities and relying less on different instance variables being set, the `get_content` method was no longer necessary, so I directly call the client here.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 4]
```diff
@@ -4,38 +4,35 @@ class GithubTag
...
+      @readme_html = fetch_readme(parsed_link)
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: I think `readme_html` is a much better name for this, the fact that it was updated (the old name was `updated_html` is really more of an implementation detail.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 5]
```diff
@@ -4,38 +4,35 @@ class GithubTag
...
 
     def render
       ActionController::Base.new.render_to_string(
         partial: PARTIAL,
         locals: {
-          content: @content,
-          show_readme: show_readme?,
-          updated_html: @updated_html
+          content: content,
+          show_readme: show_readme? && readme_html.present?,
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: If the README HTML is not present, we also skip the second part of the view.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 6]
```diff
@@ -53,10 +50,16 @@ def parse_options(link)
     end
 
     def show_readme?
-      @options.none? "no-readme"
+      options.none? "no-readme"
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: Using the `attr_accessor` instead of the instance variable.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 7]
```diff
@@ -4,38 +4,35 @@ class GithubTag
...
-          show_readme: show_readme?,
-          updated_html: @updated_html
+          content: content,
+          show_readme: show_readme? && readme_html.present?,
+          readme_html: readme_html
         },
       )
     end
 
+    private
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: Moved this up, nothing except for `render` needs to be public.

File: app/liquid_tags/github_tag/github_readme_tag.rb [hunk 8]
```diff
@@ -53,10 +50,16 @@ def parse_options(link)
...
+      options.none? "no-readme"
     end
 
-    private
+    def fetch_readme(link)
+      readme_html = client.readme(link, accept: "application/vnd.github.html")
+      readme = client.readme(link)
+      clean_relative_path!(readme_html, readme.download_url)
+    rescue Octokit::NotFound => _e
+      nil
```
Comment (citizen428) [inline app/liquid_tags/github_tag/github_readme_tag.rb]: This will set `readme_html` to `nil`, which is used in the check above.

File: app/views/liquids/_github_readme.html.erb [hunk 9]
```diff
@@ -1,7 +1,7 @@
 <div class="ltag-github-readme-tag">
   <div class="readme-overview">
     <h2>
-      <img src="<%= ActionController::Base.helpers.asset_path('github-logo.svg') %>" alt="GitHub logo">
+      <img src="<%= ActionController::Base.helpers.asset_path("github-logo.svg") %>" alt="GitHub logo">
```
Comment (citizen428) [inline app/views/liquids/_github_readme.html.erb]: Our commit hooks made me change the quotes here.

**Comment by citizen428** [comment]
If any reviewers want to play around with the code, feel free to use https://github.com/citizen428/test-dummy as a test repository.

File: app/liquid_tags/github_tag/github_readme_tag.rb (hunk 1 above)
Comment (rhymes) [reply app/liquid_tags/github_tag/github_readme_tag.rb]: they can probably be private right?

Comment (citizen428) [reply app/liquid_tags/github_tag/github_readme_tag.rb]: They can, though I often don't bother with readers. Happy to change it though if you prefer.

Comment (rhymes) [reply app/liquid_tags/github_tag/github_readme_tag.rb]: nope, it's fine :)
